tests/e2e_all/scripts/common/run_tests.sh

```diff
@@ -32,132 +27,114 @@ numClients=$(wc -w <<<"$clientVersions")
-        rm -f "$DAEMON_LOG_FRAGMENT_NAME"
-        touch "$DAEMON_LOG_FRAGMENT_NAME"
-        if ((attempts > 0)); then
-          logWarning "    Exit status was $exitStatus; will retry in 3 seconds"
-          sleep 3
-        fi
-        log "    Running test script ... "
-
-        # Execute the test script
-        timeout --foreground "$timeoutDuration" "$testScriptsDir/tests/$testToRun" "$daemonVersion" "$clientVersion" \
-          >"$stdoutFileName" 2>"$stderrFileName"
-
-        exitStatus=$?
-        if ((exitStatus != 0 && exitStatus != 50)); then
-          # shellcheck disable=SC2129
-          echo "    test execution's stdout: " | tee -a "$reportFile"
-          sed 's/^/        /' "$stdoutFileName" | tee -a "$reportFile"
-
-          echo "    test execution's stderr: " | tee -a "$reportFile"
-          sed 's/^/        /' "$stderrFileName" | tee -a "$reportFile"
-
-          echo "    daemon log fragment: " | tee -a "$reportFile"
-          sed 's/^/        /' "$DAEMON_LOG_FRAGMENT_NAME" | tee -a "$reportFile"
-
-          echo
-          echo >>"$reportFile"
-        fi
-
-        attempts=$((attempts + 1))
+      "$testScriptsDir/common/run_single_test.sh" $clientVersion $daemonVersion $testToRun $timeoutDuration
+    done
+  done
+}
+
+if [ $allowParallelization == "true" ] && command -v env_parallel >/dev/null 2>&1; then
+  logInfo "Found GNU parallel, running tests in parallel"
+  export -f run_tests_for_daemon
+  # Run a round of tests against each daemon in parallel
+  parallel --jobs 5 \
+    --env run_tests_for_daemon \
+    "testScriptsDir='$testScriptsDir' && testsToRun='$testsToRun' && clientVersions='$clientVersions' && testScriptsDir='$testScriptsDir' && timeoutDuration='$timeoutDuration' && run_tests_for_daemon" \
+    ::: $daemonVersions
+else
+  # The old way of running e2e tests - no parallelization
+  logWarning "Unable to find GNU parallel, running tests serially :("
+  for daemonVersion in $daemonVersions; do
+    for testToRun in $testsToRun; do
+      for clientVersion in $clientVersions; do
+        "$testScriptsDir/common/run_single_test.sh" $clientVersion $daemonVersion $testToRun $timeoutDuration
       done
+    done
+  done
+fi
 
-      if ((exitStatus != 0 && exitStatus != 50 && attempts == maxAttempts)); then
-        logError "    Failed after $maxAttempts attempts                                                            "
-      fi
+passed=0
+failed=0
+ignored=0
+total=0
 
-      total=$((total + 1))
-      if ((exitStatus == 0)); then
-        # Exit code 0, but did the output contain the magic 'TEST PASSED' words?
-        if ! grep -q "TEST PASSED" "$stdoutFileName"; then
-          exitStatus=51
-        fi
-      fi
+logInfo "Aggregating Test Info:"
+
+# Aggregation of all the logs at the end serially so that the final report is in order
+for daemonVersion in $daemonVersions; do
+  for testToRun in $testsToRun; do
+    for clientVersion in $clientVersions; do
```

Comment:
Aggregation of the logs synchronously so that everything has a nice order in the end